apps/web/src/app/(profile)/[username]/_components/dashboard/badges/_shared_solutions_badges.ts

```diff
@@ -0,0 +1,58 @@
+import { prisma } from '@repo/db';
+import type {AllBadgeObjs, BadgeFn} from "../_actions";
+import type {DifficultyBadges} from "~/app/(profile)/[username]/_components/dashboard/badges/_difficulty_badges";
+import type {AotBadges} from "~/app/(profile)/[username]/_components/dashboard/badges/_advent_badges";
+
+export const SolutionBadgeKeys = ['bronze', 'silver', 'gold', 'platinum'] as const;
+
+export type SolutionBadges = typeof SolutionBadgeKeys[number];
+
+export interface SharedTotals {
+  TotalCompleted: number;
+}
+
+export const sharedSolutionsBadgesFn: BadgeFn =
+  async ({
+           userId,
+           badges,
+         }: {
+    userId: string;
+    badges: AllBadgeObjs;
+  }) => {
+    // Shared Solutions Badges
+    const sharedSolutions: SharedTotals[] | null = await SharedSolutionRetrieveData(userId);
+    return await SharedBadgesFn(badges, sharedSolutions ?? []);
+  };
+
+export const SharedSolutionRetrieveData = async (userId: string) => {
+  const data: SharedTotals[] | null =
+    await prisma.$queryRaw`SELECT COUNT(SharedSolutionId) as TotalCompleted FROM (SELECT DISTINCT SharedSolutionId FROM SharedSolution JOIN Challenge ON SharedSolution.challengeId = Challenge.Id JOIN Vote ON SharedSolution.Id = Vote.SharedSolutionId WHERE SharedSolution.userId = ${userId} AND rootType = 'SHAREDSOLUTION') unique_query`;
+  return data;
+}
+
+export const AwardSolutionBadge = (slug: AotBadges | DifficultyBadges | SolutionBadges): slug is SolutionBadges => {
```

Comment:
this isn't a typeguard since youre not really validating anything. just make the return type Solution Badges

not entirely sure what this is yet, ill keep reading

Reply:
I've made a function that actually does the typeguard. This implementation of typeguards is just wrong.